biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java

```diff
@@ -228,7 +211,8 @@ private Jar buildFromInputStream(InputStream in, long lastModified) throws IOExc
 				int size = (int) entry.getSize();
 				try (ByteBufferOutputStream bbos = new ByteBufferOutputStream((size == -1) ? BUFFER_SIZE : size + 1)) {
 					bbos.write(jin);
-					putResource(entry.getName(), new EmbeddedResource(bbos.toByteBuffer(), lastModified), true);
+					long modifiedTime = ZipUtil.getModifiedTime(entry);
```

Comment:
Using zipinputstream for all zip files uses an EmbeddedResource for each zip entry which them forces the entire Jar in to memory. ZipResource will defer and possible avoid reading a zip entry into memory if the resource is never accessed. Since sometimes Jar is used to access just the manifest and maybe a few resources, forcing the entire jar into memory when it is opened is a potentially negative performance choice.

I think we need a different solution for the very few malformed jars instead of making all jars suffer.

Reply:
got it! I'll try to figure something out.